nixops/backends/ec2.py

```diff
@@ -679,97 +681,103 @@ def security_groups_to_ids(self, subnetId, groups):
+            if request.status.code in {"schedule-expired", "canceled-before-fulfillment", "bad-parameters", "system-error"}:
+                self.spot_instance_request_id = None
+                self.log_end("")
+                raise Exception("spot instance request failed with result ‘{0}’".format(request.status.code))
+            time.sleep(3)
+        self.log_end("")
 
+        instance = self._retry(lambda: self._get_instance(instance_id=request.instance_id))
+
+        return instance
 
-    def create_instance(self, defn, zone, devmap, user_data, ebs_optimized):
-        common_args = dict(
-            instance_type=defn.instance_type,
-            placement=zone,
-            key_name=defn.key_pair,
-            placement_group=defn.placement_group,
-            block_device_map=devmap,
-            user_data=user_data,
-            image_id=defn.ami,
-            ebs_optimized=ebs_optimized
-        )
 
+    def create_instance(self, defn, zone, user_data, ebs_optimized, args):
+        IamInstanceProfile = {}
         if defn.instance_profile.startswith("arn:") :
-            common_args['instance_profile_arn'] = defn.instance_profile
+            IamInstanceProfile["Arn"] = defn.instance_profile
         else:
-            common_args['instance_profile_name'] = defn.instance_profile
+            IamInstanceProfile["Name"] = defn.instance_profile
 
         if defn.subnet_id != "":
             if defn.security_groups != [] and defn.security_groups != ["default"]:
                 raise Exception("‘deployment.ec2.securityGroups’ is incompatible with ‘deployment.ec2.subnetId’")
-            common_args['network_interfaces'] = self._get_network_interfaces(defn)
+
+            args['NetworkInterfaces'] = [dict(
+                AssociatePublicIpAddress=defn.associate_public_ip_address,
+                SubnetId=defn.subnet_id,
+                DeviceIndex=0,
+                Groups=self.security_groups_to_ids(defn.subnet_id, defn.security_group_ids)
+            )]
         else:
-            common_args['security_groups'] = defn.security_groups
+            args['SecurityGroups'] = defn.security_groups
 
         if defn.spot_instance_price:
-            if self.spot_instance_request_id is None:
-
-                if defn.spot_instance_timeout:
-                    common_args['valid_until'] = \
-                        (datetime.datetime.utcnow() +
-                         datetime.timedelta(0, defn.spot_instance_timeout)).isoformat()
+            args["InstanceMarketOptions"] = dict(
+                MarketType="spot",
+                SpotOptions=dict(
+                    MaxPrice=str(defn.spot_instance_price/100.0),
+                    SpotInstanceType=defn.spot_instance_request_type,
+                    InstanceInterruptionBehavior=defn.spot_instance_interruption_behavior
+                )
+            )
+            if defn.spot_instance_timeout and defn.spot_instance_request_type != "one-time":
```

Comment:
Why do we need `defn.spot_instance_request_type != "one-time"` here?

Reply:
The argument `ValidUntil` is not relevant in a one time spot request since we cancel the request once it i fulfilled

Reply:
Yeah but I think the whole point was to cancel requests if they exceed X time without getting fulfilled.

Reply:
ok then that make sense.